code/planning/src/planning/preplanning_trajectory.py

```diff
@@ -0,0 +1,1450 @@
+                points = help_functions.linear_interpolation(
+                    start=start,
+                    end=end,
+                    interval_m=INTERVALL)
+            else:
+                radius = self.geometry_data[road_id][3][i]
+                end = help_functions.end_of_circular_arc(
+                    start_point=start, angle=hdg,
+                    length=length, radius=radius)
+                points = help_functions.circular_interpolation(
+                    start=start,
+                    end=end,
+                    arc_radius=radius)
+            for j in range(len(points)):
+                x.append(points[j][0])
+                y.append(points[j][1])
+                yaw.append(hdg)
+                speed.append(max_speed)
+        points = [x, y, yaw, speed]
+        # Delete duplicates and very close points
+        delete_index = []
+        for i in range(0, len(points[0]) - 1):
+            p = (points[0][i], points[1][i])
+            p_next = (points[0][i + 1], points[1][i + 1])
+            dist = help_functions.euclid_dist(p, p_next)
+            # print("dist", dist)
+            # point is to close to the following point (0.5m)
+            if dist < 0.5:
+                # print("SMALL")
+                delete_index.append(i)
+            # outliner point
+            elif dist > 3:
+                # print("BIG")
+                delete_index.append(i)
+        # delete the points with the calculated indices
+        number = 0
+        for i in delete_index:
+            i -= number
+            del points[0][i]
+            del points[1][i]
+            del points[2][i]
+            del points[3][i]
+            number += 1
+        return points
+
+    def get_endpoints(self, road_id: int):
+        """ Calculate the startpoint and endpoint of a given road
+            :param road_id: the road id of the examined road
+            :return: start_point, end_point
+                start_point: x and y coordinate of the starting point
+                end_point: x and y coordinate of the ending point
+        """
+        size = len(self.geometry_data[road_id][0])
+        x_start = self.geometry_data[road_id][0][0]
+        y_start = self.geometry_data[road_id][1][0]
+        start_point = (x_start, y_start)
+
+        length = self.geometry_data[road_id][4][size - 1]
+        hdg = self.geometry_data[road_id][2][size - 1]
+        # calculate the start point of the last part of the road
+        # to calculate the endpoint of the whole road
+        x = self.geometry_data[road_id][0][size - 1]
+        y = self.geometry_data[road_id][1][size - 1]
+        last_start = (x, y)
+
+        # check the last curvature value to see if it is line or arc
+        if self.geometry_data[road_id][3][size-1] == LINE:
+            xd = length * cos(hdg)
+            yd = length * sin(hdg)
+            # subtract a small value due to inaccuracy
```

Comment:
how did you obtain these numbers?

Reply:
Only for some tests. To change and comment these would be part of the refactoring